fvm/src/call_manager/default.rs

```diff
@@ -412,6 +416,198 @@ where
+        // in case it was changed (which could happen if the target upgrade endpoint
+        // sent a message to this actor which in turn called upgrade)
+        let code_id_before = state.code;
+
+        let result = self.upgrade_actor_inner::<K>(actor_id, new_code_cid, params)?;
+
+        if result.exit_code == ExitCode::OK {
+            // after running the upgrade, our code cid must not have changed
+            let code_id_after = self
+                .state_tree_mut()
+                .get_actor(actor_id)?
+                .ok_or_else(|| syscall_error!(NotFound; "actor not found: {}", actor_id))?
+                .code;
+            if code_id_before != code_id_after {
+                return Err(syscall_error!(Forbidden; "re-entrant upgrade detected").into());
+            }
+
+            self.state_tree_mut().set_actor(
+                origin,
+                ActorState::new(
+                    new_code_cid,
+                    state.state,
+                    state.balance,
+                    state.sequence,
+                    None,
+                ),
+            );
+        }
+
+        Ok(result)
+    }
+
+    fn upgrade_actor_inner<K: Kernel<CallManager = Self>>(
+        &mut self,
+        actor_id: ActorID,
+        new_code_cid: Cid,
+        params: Option<Block>,
+    ) -> Result<InvocationResult> {
+        let state = self
+            .state_tree()
+            .get_actor(actor_id)?
+            .ok_or_else(|| syscall_error!(NotFound; "actor not found: {}", actor_id))?;
+
+        let mut block_registry = BlockRegistry::new();
+
+        // add the params block to the block registry
+        let params_id = if let Some(blk) = params {
+            block_registry.put(blk)?
+        } else {
+            NO_DATA_BLOCK_ID
+        };
+
+        // also add a new block to the registry which includes the UpgradeInfo which
+        // we pass to the actor's upgrade method
+        let ui = UpgradeInfo {
+            old_code_cid: state.code,
+        };
+        let ui_params = to_vec(&ui).map_err(
+            |e| syscall_error!(IllegalArgument; "failed to serialize upgrade params: {}", e),
+        )?;
+        let ui_params_id = block_registry.put(Block::new(CBOR, ui_params))?;
+
+        log::trace!(
+            "upgrading {} from {} to {}",
+            actor_id,
+            state.code,
+            new_code_cid
+        );
+        self.map_mut(|cm| {
+            let engine = cm.engine.clone(); // reference the RC.
+
+            // Make the kernel.
+            let kernel = K::new(
+                cm,
+                block_registry,
+                actor_id,
+                actor_id,
+                1919,
+                TokenAmount::from_atto(0),
+                false,
+            );
+
+            // Make a store.
+            let mut store = engine.new_store(kernel);
+
+            // TODO: a hack until I find a better/simpler way to do this
```

Comment:
So, honestly, we can probably treat these errors as "fatal" and check earlier. I.e., the _final_ version of upgrade (not for this PR, but later) would:

1. Lookup the target code CID in some "deployed actors" table in the init actor.
2. Check some metadata generated on deploy to see if the actor supports the upgrade entrypoint.

By the time we get to _this_ code, everything will have been checked and any errors would be considered "fatal".

---

For _now_, what we'd likely do is check if:

1. The caller is one of builtin actor types X/Y/Z (in this case, it would be restricted to an eth account).
2. The target code CID is an EVM actor (again, using the builtin actors manifest).

Because that's the immediate use-case (upgrading eth accounts into EVM actors).